backend/lib/src/services/auth.rs

```diff
@@ -1,73 +1,589 @@
+            \n\
+            URI: {uri}\n\
+            Version: {version}\n\
+            Chain ID: {chain_id}\n\
+            Nonce: {nonce}\n\
+            Issued At: {issued_at}",
+        )
+    }
+
+    /// Generate a JWT token for the given address
+    fn generate_jwt(&self, address: &str) -> Result<String, Error> {
+        let now = Utc::now();
+        let exp = now + JWT_EXPIRY_OFFSET;
+
+        let claims = JwtClaims {
+            address: address.to_string(),
+            exp: exp.timestamp(),
+            iat: now.timestamp(),
+        };
+
+        jsonwebtoken::encode(&Header::default(), &claims, &self.encoding_key)
+            .map_err(|_| Error::Internal)
+    }
+
+    /// Generate a SIWE-compliant message for the user to sign
+    ///
+    /// The message will expire after a given time
+    pub async fn challenge(&self, address: &str, chain_id: u64) -> Result<NonceResponse, Error> {
+        // Validate address before generating message or storing in cache
         validate_eth_address(address)?;
 
-        Ok(NonceResponse {
-            message: format!(
-                "example.com wants you to sign in with your Ethereum account:\n{}\n\n\
-                Sign in to access your account.\n\n\
-                URI: https://example.com\n\
-                Version: 1\n\
-                Chain ID: {}\n\
-                Nonce: aBcDeF12345\n\
-                Issued At: 2024-01-01T00:00:00Z",
-                address, chain_id
-            ),
-            nonce: "aBcDeF12345".to_string(),
-        })
+        let nonce = Self::generate_random_nonce();
+        let message = Self::construct_auth_message(address, AUTH_SIWE_DOMAIN, &nonce, chain_id);
+
+        // Store message paired with address in storage
+        // Using message as key and address as value
+        self.storage
+            .store_nonce(
+                message.clone(),
+                address.to_string(),
+                AUTH_NONCE_EXPIRATION_SECONDS,
+            )
+            .await
+            .map_err(|_| Error::Internal)?;
+
+        Ok(NonceResponse { message })
     }
 
-    pub async fn verify_eth_signature(
-        &self,
-        _message: &str,
-        signature: &str,
-    ) -> Result<VerifyResponse, Error> {
-        // TODO(MOCK): check that the address matches the signature
-        if !signature.starts_with("0x") || signature.len() != 132 {
-            return Err(Error::Unauthorized("Invalid signature".to_string()));
+    fn verify_eth_signature(message: &str, signature: &str) -> Result<String, Error> {
+        let sig = PrimitiveSignature::from_str(signature)
+            .map_err(|_| Error::Unauthorized("Invalid signature format".to_string()))?;
+
+        // Hash the message with EIP-191 prefix
+        let message_hash = eip191_hash_message(message.as_bytes());
+
+        // Recover the public key from the signature
+        let recovered_pubkey = sig.recover_from_prehash(&message_hash).map_err(|_| {
```

Comment:
Recovering a pub key from the message is the same a verifing it?
Also, are you using this at all?